File: src/main/java/com/github/kpavlov/jreactive8583/netty/pipeline/ParseExceptionHandler.java
package com.github.kpavlov.jreactive8583.netty.pipeline;

import com.github.kpavlov.jreactive8583.iso.MessageFactory;
import com.solab.iso8583.IsoMessage;
import com.solab.iso8583.IsoType;
import io.netty.channel.ChannelHandler;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;

import javax.annotation.Nonnull;
import java.text.ParseException;

/**
 * Handles {@link ParseException}s and responds with administrative message
 *
 * @see <a href="http://stackoverflow.com/questions/28275677/how-to-answer-an-invalid-iso8583-message">StackOverflow: How to answer an invalid ISO8583 message</a>
 */
@ChannelHandler.Sharable
public class ParseExceptionHandler extends ChannelInboundHandlerAdapter {

    private final MessageFactory<IsoMessage> isoMessageFactory;

    private final boolean includeErrorDetails;

    public ParseExceptionHandler(@Nonnull MessageFactory<IsoMessage> isoMessageFactory,
                                 boolean includeErrorDetails) {
        this.isoMessageFactory = isoMessageFactory;
        this.includeErrorDetails = includeErrorDetails;
    }

    @Override
    public void exceptionCaught(@Nonnull ChannelHandlerContext ctx,
                                @Nonnull Throwable cause) throws Exception {
        if (cause instanceof ParseException) {
            final IsoMessage message = createErrorResponseMessage((ParseException) cause);
            ctx.writeAndFlush(message);
        }
        super.exceptionCaught(ctx, cause);
    }

    protected IsoMessage createErrorResponseMessage(ParseException cause) {
        final IsoMessage message = isoMessageFactory.newMessage(0x1644);
        message.setValue(24, 650, IsoType.NUMERIC, 3); //650 (Unable to parse message)
        if (includeErrorDetails) {
            String details = cause.getMessage();
            if (details.length() > 25) {
                details = details.substring(0, 22) + "...";
            }
            message.setValue(44, details, IsoType.LLVAR, 25);
        }
        return message;
    }
}

File: src/main/kotlin/com/github/kpavlov/jreactive8583/netty/pipeline/ParseExceptionHandler.kt
package com.github.kpavlov.jreactive8583.netty.pipeline

import com.github.kpavlov.jreactive8583.iso.MessageClass
import com.github.kpavlov.jreactive8583.iso.MessageFactory
import com.github.kpavlov.jreactive8583.iso.MessageFunction
import com.github.kpavlov.jreactive8583.iso.MessageOrigin
import com.solab.iso8583.IsoMessage
import com.solab.iso8583.IsoType
import io.netty.channel.ChannelHandler.Sharable
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import java.text.ParseException

/**
 * Handles [ParseException]s and responds with administrative message
 *
 * @see [StackOverflow: How to answer an invalid ISO8583 message](http://stackoverflow.com/questions/28275677/how-to-answer-an-invalid-iso8583-message)
 */
@Sharable
public open class ParseExceptionHandler(
    private val isoMessageFactory: MessageFactory<IsoMessage>,
    private val includeErrorDetails: Boolean,
) : ChannelInboundHandlerAdapter() {
    @Deprecated("Deprecated in Java")
    @Throws(Exception::class)
    public override fun exceptionCaught(
        ctx: ChannelHandlerContext,
        cause: Throwable,
    ) {
        if (cause is ParseException) {
            val message = createErrorResponseMessage(cause)
            ctx.writeAndFlush(message)
        }
        ctx.fireExceptionCaught(cause)
    }

    @Suppress("MagicNumber")
    protected fun createErrorResponseMessage(cause: ParseException): IsoMessage {
        val message =
            isoMessageFactory.newMessage(
                MessageClass.ADMINISTRATIVE,
                MessageFunction.NOTIFICATION,
                MessageOrigin.OTHER,
            )
        // 650 (Unable to parse message)
        message.setValue(24, 650, IsoType.NUMERIC, 3)
        if (includeErrorDetails) {
            var details = cause.message
            if (details != null && details.length > 25) {
                details = details.take(22) + "..."
            }
            message.setValue(44, details, IsoType.LLVAR, 25)
        }
        return message
    }
}
